### ISSUE TITLE
Should service action (restore) remove all node ports from a service?

### ISSUE THREAD

**Issue by ncdc** [heptio/ark#354]
We are currently removing all nodePorts from all services when we restore them:

https://github.com/heptio/ark/blob/973f630cc7b2bdd464cc4e78d776ab4099a43381/pkg/restore/service_action.go#L60-L63

This doesn't really make sense to me. @skriss do you remember what the original motivation for this was?

**Comment by skriss** [comment]
IIRC it was because keeping them makes the "clone namespace" use case fail due to port collision.

**Comment by ncdc** [comment]
Ok, that makes sense. I don't think they should generally be dropped by default, though. We should think through ways to allow the user to express the desired behavior in a clean manner.

**Comment by ncdc** [comment]
cc @jbeda - another UX question. How do we support both 1) restore into new cluster and 2) restore into new cluster into a different namespace?

**Comment by jbeda** [comment]
My answer would be similar to https://github.com/heptio/ark/issues/192#issuecomment-371653916.  By all means let's come up with granular flags for stuff like this. Or have a config file for specifying what gets remapped how.  But then package up the common use cases under a single flag.  Document it so that users can tweak it if they want.

**Comment by ncdc** [comment]
Makes sense. Maybe we allow users to remap node ports as an option?
```
ark restore create --from-backup foo --namespace-mappings bob1:bob2 --node-port-mappings 40001:40002,41234:42345
```

**Comment by jbeda** [comment]
That sounds great.  But from a user point of view you'd want to be able to ask "what node ports does this backup/namespace use?" and "Will those conflict with the cluster I'm trying to apply to?"  Answering those questions doesn't fit cleanly into this model unfortunately.

In addition, you may want to let the system pick the node port as you remap.  Unfortunately, it is difficult to know if the node port in the backup was assigned by the system during admission control or if it was assigned explicitly by the user.  Once Daniel's big multi-merge thing goes in we may have more data.  To really tell what the intent was we'd have to start cracking into the annotation added by `kubectl apply`.

Perhaps we need some sort of a restore dry run that helps to suss out these issues and can give you a summary of the resources to be created along with perhaps testing those against the target cluster?  That is a pretty big new feature though.

**Comment by ncdc** [comment]
Yeah, this is potentially a pretty big thing... Right now it's "fire and forget it" when it comes to restoring. The server does the best it can, but it's all automated and there's no opportunity for user input. Nor do we have a holistic **in-memoryr** view of either the backed up resources or the destination cluster's resources, so we can't currently easily evaluate everything and say "hey, your node port is going to conflict".

**Comment by rosskukulinski** [comment]
Ref #448 

This is tricky, because NodePort can be defined by the user, or can be selected by kube-controller-manager at creation time.  IMO, if kube-controller-manager picked the NodePort, then let that process happen again.  If the user specified it, then we probably need a remap mechanism.  I don’t believe this information is actually available to Ark after the fact.

This is a big-impact issue that touches UX (via CLI flags) and user expectations.  Will need more thought around use-cases.

**Comment by jbeda** [comment]
This is tricky but i think there is a generally accepted way here.  There are a set of annotations that list the *specified* state of the resource.  `kubectl apply` uses these.  From here you can figure out user intent.  it isn't 100% guaranteed to work but it is something that we can document.

I'd say:
1. If the annotation is there and the user specified the node port then keep that specification.  if they didn't then remove it from the backup.
2. If the annotation is missing, then default to *something* and give users an option to modify that default.  I'm leaning toward removing the node port specification but could go either way.

Over time, we may need a way to specify generic munging policy so that users can specify how/where to remove stuff like this on a per-restore basis.

Also be aware that there is work to make that annotation more official with a feature called "server side apply" and that we should be tracking that.  BryanL has experience with this issue wrt ksonnet  and it may be worth reaching out to him.

**Comment by rosskukulinski** [comment]
WFM +1

Perhaps #474 could help in terms of generic munging policy down the road.

**Comment by timoreimann** [comment]
Just to add one user data point: We just encountered this very issue when running a test recovery with Ark on one of our namespaces, where the omission of the node ports eventually prevented some access patterns we've employed before. In this very case, it'd have been desirable to default to keeping the specified node ports, or at least have an option to enforce such behavior.

It's nothing overly critical for us right now though as we pin node ports in a single, non-production use case only. However, it's something we'll have to look out for when creating new service specs.

The approach that @jbeda describes to leverage the specified JSON state from the annotation (when available) sounds like a reasonable first step. If that's something we could agree to, I'd be happy to take a stab at a PR. Please let me know.

**Comment by rosskukulinski** [comment]
Thanks for volunteering, @timoreimann!  @ncdc / @skriss does Joe's approach work for you? [My suggestion for (2) is to default to removing nodePort and add CLI flag to enable keeping it]

**Comment by skriss** [comment]
It makes sense to me, and I think removing it by default is sensible.

**Comment by skriss** [comment]
this was fixed by #712, closing.
